Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs

```diff
@@ -0,0 +1,522 @@
+                    {
+                        orderBook.UpdateBidRow(price, amount);
+                    }
+                    else if (amount < 0)
+                    {
+                        orderBook.UpdateAskRow(price, amount);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        private void OnOrderClose(string[] entries)
+        {
+            string brokerId = entries[0];
+            if (entries[5].IndexOf("canceled", StringComparison.OrdinalIgnoreCase) >= 0)
+            {
+                var order = CachedOrderIDs
+                    .FirstOrDefault(o => o.Value.BrokerId.Contains(brokerId))
+                    .Value;
+                if (order == null)
+                {
+                    order = _algorithm.Transactions.GetOrderByBrokerageId(brokerId);
+                    if (order == null)
+                    {
+                        // not our order, nothing else to do here
+                        return;
+                    }
+                }
+                Order outOrder;
+                if (CachedOrderIDs.TryRemove(order.Id, out outOrder))
+                {
+                    OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, 0, "Bitfinex Order Event") { Status = OrderStatus.Canceled });
+                }
+            }
+            else
+            {
+                lock (closeLocker)
+                {
+                    _pendingClose.Add(brokerId);
+                }
+            }
+        }
+
+        private void OnTradeExecuted(string[] entries)
+        {
+            try
+            {
+                var brokerId = entries[4];
+                var order = CachedOrderIDs
+                    .FirstOrDefault(o => o.Value.BrokerId.Contains(brokerId))
+                    .Value;
+                if (order == null)
+                {
+                    order = _algorithm.Transactions.GetOrderByBrokerageId(brokerId);
+                    if (order == null)
+                    {
+                        // not our order, nothing else to do here
+                        return;
+                    }
+                }
+                Trade trade = Trade.FromTU(entries);
+                EmitTradeTick(order, trade);
+                EmitFillOrder(order, trade);
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        private void EmitTradeTick(Order order, Trade trade)
+        {
+            lock (TickLocker)
+            {
+                Ticks.Add(new Tick
+                {
+                    Value = trade.Price,
+                    Time = trade.Time,
+                    Symbol = _symbolMapper.GetLeanSymbol(trade.Pair),
+                    TickType = TickType.Trade,
+                    Quantity = trade.Amount
```

Comment:
As much as I would love to have _signed quantity_ in Trade ticks, LEAN does not support this for now (volume based indicators currently assume positive values), we would probably need to add the `TradeDirection` to the `Tick` class.
For now I recommend we use `Math.Abs(trade.Amount)` here.

Reply:
Okay